source/common/rds/rds_route_config_subscription.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "envoy/config/core/v3/config_source.pb.h"
+#include "envoy/config/subscription.h"
+#include "envoy/rds/route_config_provider.h"
+#include "envoy/rds/route_config_update_receiver.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/stats/scope.h"
+
+#include "source/common/grpc/common.h"
+#include "source/common/init/manager_impl.h"
+#include "source/common/init/target_impl.h"
+#include "source/common/init/watcher_impl.h"
+#include "source/common/rds/route_config_provider_manager.h"
+
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Rds {
+
+/**
+ * All RDS stats. @see stats_macros.h
+ */
+#define ALL_RDS_STATS(COUNTER, GAUGE)                                                              \
+  COUNTER(config_reload)                                                                           \
+  COUNTER(update_empty)                                                                            \
+  GAUGE(config_reload_time_ms, NeverImport)
+
+/**
+ * Struct definition for all RDS stats. @see stats_macros.h
+ */
+struct RdsStats {
+  ALL_RDS_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+/**
+ * A class that fetches the route configuration dynamically using the RDS API and updates them to
+ * RDS config providers.
+ */
+class RdsRouteConfigSubscription : Envoy::Config::SubscriptionCallbacks,
```

Comment:
@adisuissa I tend to agree, but I think `SubscriptionBase` is very minimal and has been stable a long time. One pattern you could do is to have this not inherit `SubscriptionCallbacks`, define methods with `onConfigUpdate`-like signatures, and then for each RDS-type add a veneer that does inherit from `SubscriptionBase` and whether the `onConfigUpdate` just points at the `onConfigUpdate`-like methods. @tkovacs-2 would this make sense? How much work do you see this as being? Should it be a followup or would it make sense here?

In the interest of getting this merged and moving forward, I'd propose we allow this and take it as a followup to try and avoid

Reply:
I think understand the suggestion, but I think it is not good idea.

The `SubscriptionBase` (today) is quite thin. Just the `resource_decorer_` member and a wrapper function to fetch the resource type name. We can say no any logic here. Yes, in the future this may change, but may be not.

On the other hand if we add this veneer class we have to solve problems:
- If we keep it as template that is against the lot of effort we did to eliminate templates from `rds` folder. If we say all rds type have to implement separately that is duplication.
- How does the `Rds::RdsRouteConfigSubscription` refer to the veneer? It has to somehow to reuse the `resource_decoder_` object. We should add some new interface, but that is just increases complexity.
If we don't want this and rather create the `subscription_` in the veneer then we have to move other stuff a well, like all the `init`-s, `scope_`, etc.
- Related to the previous, which object owns which?

Of course everything is possible, the question is the gain and the price. I think this doesn't worth.

Reply:
I'm fine with this!
@htuch please take a final look.

Reply:
Yes, I agree, I'm not a huge fan of lots of complexity if we can avoid. Thanks for taking a look.